### ISSUE TITLE
Create a virtual memory manager for the x86_64 ISA

### ISSUE THREAD

**Issue by mdpcs3544** [CharlotteOS-Deprecated/charlotte-core#11]
Create a virtual memory manager for the x86_64 ISA that can do the following:

- [x] ~~Enable the use of PCIDs~~*
- [x] Create a page map
- [x] ~~Implement `Clone` and `Drop` for page map~~*
- [x] Load a page map
- [x] Map normal 4 KiB pages
- [x] Unmap normal pages
- [x] Map large pages
- [x] Unmap large pages
- [x] Detect CPU support for huge pages
- [x] Map huge pages
- [x] Unmap huge pages
- [x] ~~Handle page fault exceptions~~**
- [x] ~~Make use of the page attribute table (PAT) to to set memory attributes for different regions~~***

*Postponed until there are multiple PageMaps
**Postponed until there is a userspace
***Moved to a separate issue due to the possibility of breaking the bootloader provided PageMap

**Comment by mdpatelcsecon** [comment]
This will need to be pushed back as the PMM need to be rewritten and the design for access control needs to be agreed upon.

**Comment by mdpcs3544** [comment]
This has been unblocked with the completion of the initial version of the PMM.

**Comment by mdpcs3544** [comment]
Blocked on #76

**Comment by mdpcs3544** [comment]
#76 has been declared invalid so this is no longer blocked.

**Comment by mdpcs3544** [comment]
Most of the hard parts of the VMM have been implemented. Once the unmap page function is complete then mapping and unmapping large and huge pages should be simple given that the same code used for normal pages can easily be modified to perform those operations as well. Things already build without errors however only testing will determine if the desired behavior has been implemented correctly.

**Comment by mdpcs3544** [comment]
The VMM is complete and should work for just the kernel PageMap for now. The PAT has not been modified however doing so will be a separate issue.
